controllers/logicalvolume_controller.go

```diff
@@ -273,10 +274,15 @@ func (r *LogicalVolumeReconciler) createLV(ctx context.Context, log logr.Logger,
 }
 
 func (r *LogicalVolumeReconciler) expandLV(ctx context.Context, log logr.Logger, lv *topolvmv1.LogicalVolume) error {
-	// lv.Status.CurrentSize is added in v0.4.0 and filled by topolvm-controller when resizing is triggered.
-	// The reconciliation loop of LogicalVolume may call expandLV before resizing is triggered.
-	// So, lv.Status.CurrentSize could be nil here.
+	// WA: since Status.CurrentSize is added in v0.4.0, it may be missing.
+	// To ensure the field existed, update it here.
+	reqBytes := lv.Spec.Size.Value()
 	if lv.Status.CurrentSize == nil {
+		lv.Status.CurrentSize = resource.NewQuantity(reqBytes, resource.BinarySI)
```

Comment:
I think you cannot use `reqBytes` here.

The value of `reqBytes` comes from `lv.Spec.Size.Value()`. This is set to the expanded capacity at [(s *LogicalVolumeService) ExpandVolume](https://github.com/topolvm/topolvm/blob/6b74e7f1d02ad141ae0aaa683bb804f7bc5785f5/driver/internal/k8s/logicalvolume_service.go#L310) before `expandLV` is called.

If we assign `reqBytes` to `lv.Status.CurrentSize` here, the status is updated in the subsequent line despite that the actual LV size has not been expanded.

Then, in the next reconciliation which is incurred by the status update, the following condition holds true, and the volume will never be expanded.
https://github.com/topolvm/topolvm/blob/6b74e7f1d02ad141ae0aaa683bb804f7bc5785f5/controllers/logicalvolume_controller.go#L289

IMO, when `lv.Status.CurrentSize` is nil, its value should be set **before the spec is updated**. Though you deleted the following lines in this PR, I guess they are necessary.
https://github.com/topolvm/topolvm/blob/af937939f2f984da31bc6977c9909b54a8579647/driver/controller.go#L593-L597

Reply:
Good catch! It is true that status fields should be updated after reconciliation succeeds for retrying. Your case that the status is updated before the spec is updated means reconciliation succeeds and the status update happens after the reconciliation. Therefore, we can move the status update from topolvm-controller to topolvm-node.

To move the update after resizing is completed, we ensure lvmd can accept expanding a volume to the original size. Happily, it is already done. So I'll just update the code to continue the resize logic same as actual resizing if CurrentSize is nil. Also, I'll add test cases for this.